tests/test_dataclasses.py

```diff
@@ -77,3 +93,476 @@ def test_from_dict():
+        )
+        bid_offer.update_price(30)
+        assert bid_offer.price == 30
+        assert bid_offer.energy_rate == 30 / bid_offer.energy
+
+    def test_update_energy(self):
+        bid_offer = BaseBidOffer(
+            **self.initial_data
+        )
+        bid_offer.update_energy(40)
+        assert bid_offer.energy == 40
+        assert bid_offer.energy_rate == bid_offer.price / 40
+
+    def test_to_json_string(self):
+        bid_offer = BaseBidOffer(
+            **self.initial_data
+        )
+        obj_dict = deepcopy(bid_offer.__dict__)
+
+        obj_dict["type"] = "BaseBidOffer"
+        assert bid_offer.to_json_string() == json.dumps(obj_dict, default=my_converter)
+        assert json.loads(
+            bid_offer.to_json_string(my_extra_key=10)).get("my_extra_key") == 10
+
+    def test_serializable_dict(self):
+        bid_offer = BaseBidOffer(
+            **self.initial_data
+        )
+        assert bid_offer.serializable_dict() == {
+            "type": "BaseBidOffer",
+            "id": str(bid_offer.id),
+            "energy": bid_offer.energy,
+            "energy_rate": bid_offer.energy_rate,
+            "original_price": bid_offer.original_price,
+            "time": datetime_to_string_incl_seconds(bid_offer.time),
+            "attributes": bid_offer.attributes,
+            "requirements": bid_offer.requirements
+        }
+
+    def test_from_json(self):
+        offer = Offer(
+            **self.initial_data,
+            seller="seller"
+        )
+        offer_json = offer.to_json_string()
+        assert offer == BaseBidOffer.from_json(offer_json)
+
+        bid = Bid(
+            **self.initial_data,
+            buyer="buyer"
+        )
+        bid_json = bid.to_json_string()
+        assert bid == BaseBidOffer.from_json(bid_json)
+
+
+class TestOffer:
+    def setup_method(self):
+        self.initial_data = {
+            "id": uuid.uuid4(),
+            "time": DateTime.now(),
+            "price": 10,
+            "energy": 30,
+            "original_price": 8,
+            "attributes": {},
+            "requirements": [],
+            "seller": "seller"
+        }
+
+    def test_init(self):
+        offer = Offer(
+            **self.initial_data
+        )
+        assert offer.id == str(self.initial_data["id"])
+        assert offer.time == self.initial_data["time"]
+        assert offer.price == self.initial_data["price"]
+        assert offer.energy == self.initial_data["energy"]
+        assert offer.original_price == self.initial_data["original_price"]
+        assert offer.energy_rate == self.initial_data["price"] / self.initial_data["energy"]
+        assert offer.attributes == self.initial_data["attributes"]
+        assert offer.requirements == self.initial_data["requirements"]
+        assert offer.seller == "seller"
+        assert offer.seller_id is None
+        assert offer.seller_origin is None
+        assert offer.seller_origin_id is None
```

Comment:
These should be a non None

Reply:
https://github.com/gridsingularity/d3a-interface/pull/253/files?file-filters%5B%5D=.py&hide-deleted-files=true#r708031409